core/src/test/java/io/grpc/internal/PickFirstLeafLoadBalancerTest.java

```diff
@@ -0,0 +1,2577 @@
+    loadBalancer.acceptResolvedAddresses(
+        ResolvedAddresses.newBuilder().setAddresses(newServers).setAttributes(affinity).build());
+    assertEquals(CONNECTING, loadBalancer.getCurrentState());
+    inOrder.verify(mockHelper).createSubchannel(createArgsCaptor.capture());
+    // corresponds to socketAddr2
+    inOrder.verify(mockSubchannel3).start(stateListenerCaptor.capture());
+    SubchannelStateListener stateListener3 = stateListenerCaptor.getValue();
+    // corresponds to socketAddr3
+    inOrder.verify(mockSubchannel4).start(stateListenerCaptor.capture());
+    SubchannelStateListener stateListener4 = stateListenerCaptor.getValue();
+    inOrder.verify(mockSubchannel1).shutdown();
+    inOrder.verify(mockHelper).updateBalancingState(eq(CONNECTING), pickerCaptor.capture());
+    inOrder.verify(mockSubchannel3).requestConnection();
+
+    // If obselete subchannel becomes ready, the state should not be affected
+    stateListener.onSubchannelState(ConnectivityStateInfo.forNonError(READY));
+    stateListener2.onSubchannelState(ConnectivityStateInfo.forNonError(READY));
+    assertEquals(CONNECTING, loadBalancer.getCurrentState());
+
+    // Request a connection via pickSubchannel()
+    picker = pickerCaptor.getValue();
+
+    // Calling pickSubchannel() requests a connection, gives the same result when called twice.
+    assertEquals(picker.pickSubchannel(mockArgs), picker.pickSubchannel(mockArgs));
+
+    // Fail connection attempt to address 2 (subchannel3)
+    Status error = Status.UNAVAILABLE.withDescription("Simulated connection error");
+    stateListener3.onSubchannelState(ConnectivityStateInfo.forTransientFailure(error));
+    assertEquals(CONNECTING, loadBalancer.getCurrentState());
+
+    // Starting connection attempt to address 3 (subchannel4)
+    assertEquals(CONNECTING, loadBalancer.getCurrentState());
+    inOrder.verify(mockSubchannel4).requestConnection();
+    assertEquals(CONNECTING, loadBalancer.getCurrentState());
+
+    // Connection attempt to address 2 is successful
+    stateListener4.onSubchannelState(ConnectivityStateInfo.forNonError(READY));
+    assertEquals(READY, loadBalancer.getCurrentState());
+
+    // Successful attempt shuts down other subchannels
+    inOrder.verify(mockHelper).updateBalancingState(eq(READY), pickerCaptor.capture());
+    picker = pickerCaptor.getValue();
+    inOrder.verify(mockSubchannel3).shutdown();
+
+    // Verify that picker returns correct subchannel
+    assertEquals(PickResult.withSubchannel(mockSubchannel4), picker.pickSubchannel(mockArgs));
+    assertNotEquals(PickResult.withSubchannel(mockSubchannel1), picker.pickSubchannel(mockArgs));
+    assertNotEquals(PickResult.withSubchannel(mockSubchannel2), picker.pickSubchannel(mockArgs));
+    assertNotEquals(PickResult.withSubchannel(mockSubchannel2), picker.pickSubchannel(mockArgs));
+  }
+
+  @Test
+  public void updateAddresses_overlapping_on_others_transient_failure() {
```

Comment:
They are treated slightly differently in the code--intersecting refers to when the current address that we are looking at is present in the new list while overlapping refers to overlapping addresses that are not at the current address. They behave a bit differently given different states so I think it may be good to keep them.

Reply:
I see in one test: (0,1) updated to (1,2)
another one: (0, 1) updated to (0,2)
so can we combine these to (0,1,2) updated to (0,2,3), or other  combinations, e.g. flips the order.
Major concern here is that tests are too long and hard to maintain. Let's think of ways to shorted them but with similar test coverage. FYI in round robin, test address update only in one test case and I do feel its sufficient https://github.com/grpc/grpc-java/blob/master/util/src/test/java/io/grpc/util/RoundRobinLoadBalancerTest.java#L180